plutus-core/plutus-core/src/PlutusCore/Default/Builtins.hs

```diff
@@ -740,6 +737,97 @@ list [1] and that the type tag of the element being prepended equals the type ta
 the list [2] (extracted from the type tag for the whole list constant [1]).
 -}
 
+{- Note [Builtins and Plutus type checking]
+There's a direct correspondence between the Haskell type of the denotation of a builtin and the
+Plutus type of the builtin:
+
+1. elaboration turns a Haskell type variable into a concrete Haskell type representing a Plutus type
+   variable, which later becomes demoted to a regular Haskell value representing a Plutus type
+   variable (as a part of the AST)
+2. a builtin head (i.e. a completely uninstantiated built-in type such as @Bool@ and @[]@) is
+   considered abstract by the Plutus type checker. All the type checker cares about is being able to
+   get the (Plutus) kind of a builtin head and check two builtin heads for equality
+3. Plutus type normalization tears partially or fully instantiated built-in types (such as
+   @[Integer]@) apart and creates a Plutus type application for each Haskell type application
+4. 'Emitter' and 'EvaluationResult' do not appear on the Plutus side, since the logging and failure
+   effects are implicit in Plutus as was discussed above
+5. 'Opaque' and 'SomeConstant' both carry a Haskell @rep@ type argument representing some Plutus
+   type to be used for Plutus type checking
+
+This last part means that one can attach any (legal) @rep@ to an 'Opaque' or 'SomeConstant' and
+it'll be used by the Plutus type checker completely regardless of what the built-in function
+actually does. Let's look at some examples.
+
+1. The following built-in function unlifts a 'Bool' and returns it back:
+
+    toBuiltinMeaning IdIntegerAsBool =
+        makeBuiltinMeaning
+            idIntegerAsBool
+            <costingFunction>
+      where
+        idIntegerAsBool :: SomeConstant uni Integer -> EvaluationResult (SomeConstant uni Integer)
+        idIntegerAsBool = \case
+            con@(SomeConstant (Some (ValueOf DefaultUniBool _))) -> EvaluationSuccess con
+            _                                                    -> EvaluationFailure
+
+but on the Plutus side its type is
+
+    integer -> integer
+
+because the @rep@ that 'SomeConstant' carries is 'Integer' in both the cases (in the type of the
+argument, as well as in the type of the result).
+
+This means that for this built-in function the Plutus type checker will accept a program that fails
```

Comment:
> Plutus type checker will accept a program that fails at runtime

This is already the case for all builtin functions, right? Like many built-in functions expect constant arguments, and will fail at runtime if given a non-constant.